Fix resume_from_checkpoint for deepspeed

mosheber opened the pull request:
@stas00  This fixes the resume_from_checkpoint for deepspeed, by ensuring that the deepspeed engine is the one to load the checkpoint.

# What does this PR do?

It disables the regular load_from_checkpoint, and allowing it to go to the deepspeed engine instead.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

mosheber committed:
Fix resume_from_checkpoint for deepspeed

Fix resume_from_checkpoint for deepspeed, by ensuring that the deepspeed engine is the one to load the checkpoint.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21735). All of your documentation changes will be reflected on that endpoint.

stas00 commented on src/transformers/trainer.py:
```diff
@@ -1616,7 +1616,7 @@ def train(
                 raise ValueError(f"No valid checkpoint found in output directory ({args.output_dir})")
 
-        if resume_from_checkpoint is not None and not is_sagemaker_mp_enabled():
+        if resume_from_checkpoint is not None and not is_sagemaker_mp_enabled() and args.deepspeed is None:
```
It's already skipped inside here:

https://github.com/huggingface/transformers/blob/619d51e01f326d298687912d1f65f8d460f2a6e2/src/transformers/trainer.py#L2090-L2092

Does it fail for you in some preliminary checks that happen before the code I quoted above?

stas00 replied:
Perhaps let's change line 1619, to just:
```
if resume_from_checkpoint is not None:
```
and start `self._load_from_checkpoint` with:

```
    def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
        if is_sagemaker_mp_enabled() or args.deepspeed:
            # sagemaker gets resumes in ...
            # deepspeed gets resumes in deepspeed_init
            return
```
and remove the code I quoted in the previous comment?

stas00 replied:
or keep the version you proposed and then delete the code I quoted 2 comments up.

Let's ask @sgugger since he would be the final judge anyway. Sylvain is it better to quickly exit from the function as I proposed or not enter it like Moshe proposed?

In either way we should make things consistent as they aren't now. (sagemaker does the 2nd way)

mosheber replied:
@stas00 I personally think that simply skipping the _load_from_checkpoint method if deepspeed is used, is the most straightforward way to fix this issue.

Its only a single if statement, and deepspeed_init already takes care of resuming the checkpoint correctly, the only thing that the train() method needs to do, is to let deepspeed do its own thing.

As for consistency, each method (regular, sagemaker, deepspeed) resumes the checkpoint using different logic and separate functions, thus changing that could potentially introduce more bugs into the process.

mosheber replied:
As for the error itself, it is raised since deepspeed checkpoints do not contain WEIGHTS_NAME = "pytorch_model.bin":

https://github.com/huggingface/transformers/blob/aa3787c8f0b1ee4d4fa74fe890a39cfeb601d5e5/src/transformers/trainer.py#L2073

stas00 replied:
yes, that's what I was suspecting - I wonder why nobody reported it since ever. Thank you for that!

--------------------

ok, if you prefer it this way (and let's assume Sylvain would agree) in that case then please adapt your PR to remove this code 

https://github.com/huggingface/transformers/blob/619d51e01f326d298687912d1f65f8d460f2a6e2/src/transformers/trainer.py#L2090-L2092

and replacing the next line's `elif` with `if`

stas00 replied:
and please see the next comment about CircleCI - apparently we have no control over it - so you have to go to CircleCI - logout, login and that might do the trick. I have no idea why your CircleCI creds have anything to do with our project. Sounds like a bug in CircleCI. Thanks.

ydshieh replied:
The CI is triggered and green. But @stas00 above suggestion (about removing a block) is not applied yet. I will let @mosheber and @stas00 to continue, and jump in  in case CI has further issue.

ydshieh commented:
Hi @mosheber! The CI is not triggered. It seems there is an issue with your CircleCI permissions, the tests won't run.
Could you first try refreshing your permissions as shown [here](https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions-)?

Thank you

sgugger reviewed (approved):
I found the solution of exiting early more elegant, but this works too. Thanks!

mosheber commented:
> 

@ydshieh @sgugger , thank you for approving! For some reason the CircleCI still wont run properly, I tried logging out, revoking, logging back in, refreshing, yet to avail. Could you perform the merge on your end? Or perhaps trigger the CI run?

ydshieh commented:
Hi @mosheber Thank you for trying, and sorry about this. I will push an empty commit to your PR branch to trigger CI - are you OK with it?

ydshieh committed:
Empty commit to trigger CI

ydshieh commented:
Well, I think one more step to follow:


Could you check if you are following huggingface/transformers instead of your own fork. You can check it at this link
https://app.circleci.com/projects/project-dashboard/github/mosheber/

If you are following your own fork, you have to unfollow it, and follow `huggingface/transformers` instead.



```
If a user submits a pull request to your repository from a fork, but no pipeline is triggered, then the user most likely is following a project fork on their personal account rather than the project itself of CircleCI, causing the jobs to trigger under the user’s personal account and not the organization account. To resolve this issue, have the user unfollow their fork of the project on CircleCI and instead follow the source project. This will trigger their jobs to run under the organization when they submit pull requests.
```

https://circleci.com/docs/oss/#build-pull-requests-from-forked-repositories

ydshieh commented:
If you are OK, I can also fork your PR, create a new one, but add your name as a contributor then merge the new PR. This might be easier.

mosheber commented:
> If you are OK, I can also fork your PR, create a new one, but add your name as a contributor then merge the new PR. This might be easier.

This seems to be the easiest approach, lets go with that

ydshieh commented:
well, while I pushed to `huggingface/transformers`, the CI here is triggered ... let's see.

mosheber commented:
> well, while I pushed to `huggingface/transformers`, the CI here is triggered ... let's see.

Looks like all checks have passed, thanks! Merging should now be possible

ydshieh commented:
> > well, while I pushed to `huggingface/transformers`, the CI here is triggered ... let's see.
> 
> Looks like all checks have passed, thanks! Merging should now be possible

Yeah, but see my comment

https://github.com/huggingface/transformers/pull/21735#discussion_r1117246877

stas00 commented:
> Looks like all checks have passed, thanks! Merging should now be possible

But it's not ready. Please revisit: https://github.com/huggingface/transformers/pull/21735#discussion_r1116267795

mosheber committed:
Removed deepspeed skipping 

Removed deepspeed skipping inside the _load_from_checkpoint function, as it is obsolete

mosheber commented:
> > Looks like all checks have passed, thanks! Merging should now be possible
> 
> But it's not ready. Please revisit: [#21735 (comment)](https://github.com/huggingface/transformers/pull/21735#discussion_r1116267795)

Sure thing, I removed it and changed the elif

stas00 committed:
Merge remote-tracking branch 'origin/main' into mosheber/main

stas00 commented:
Thanks Moshe, I need to run offline tests since deepspeed tests don't run on live CI (need gpus), and will merge once all is green there.

stas00 committed:
another adjustment

stas00 commented:
@ydshieh - do you by chance have any idea why CI isn't running? This time it appears to be some other problem than the original one we discussed in this issue. Thank you!

stas00 committed:
Trigger CI

stas00 commented:
Moshe, CircleCI doesn't like something about your CircleCI account. Since there is nothing I can do about it and I don't want this to drag forever I've recreated your PR here https://github.com/huggingface/transformers/pull/21798 - so let's finish it there. Thank you.

mosheber committed:
trigger circleci

mosheber commented:
> Moshe, CircleCI doesn't like something about your CircleCI account settings. Since there is nothing I can do about it and I don't want this to drag forever I've recreated your PR here #21798 - so let's finish it there. Thank you.

No problem. Just in case, I also tried to trigger another CI run after unfollowing the project as suggested here:

https://support.circleci.com/hc/en-us/articles/360008097173-Troubleshooting-why-pull-requests-are-not-triggering-jobs-on-my-organization-

Maybe it will help too

stas00 commented:
Thank you for digging this up, Moshe. The relevant quote seems to be:

> If you're following the fork instead of the upstream repo
> 
> A user who submits a pull request to your repository from a fork, but no pipeline is triggered with the pull request. This can happen when the user is following the project fork on their personal account rather than the project itself on CircleCI.
> 
> This will cause the jobs to trigger under the user's personal account. If the user is following a fork of the repository on CircleCI, we will only build on that fork and not the parent, so the parent’s PR will not get status updates. 
> 
> In these cases, the user unfollows their fork of the project on CircleCI. This will trigger their jobs to run under the organization when they submit pull requests. Those users can optionally follow the source project if they wish to see the pipelines.

which as you said you did and the CI has started. Excellent work - now we would know what to tell the users in the future.

stas00 committed:
style

mosheber commented:
> 

Glad I could help! Thanks!
